server/src-lib/Hasura/Server/Migrate.hs

```diff
@@ -84,6 +84,46 @@ data MigrationPair m = MigrationPair
   , mpDown :: Maybe (m ())
   }
 
+doesSchemaExist :: (MonadTx m) => SchemaName -> m Bool
+doesSchemaExist schemaName =
+  liftTx $ (runIdentity . Q.getRow) <$> Q.withQE defaultTxErrorHandler [Q.sql|
+    SELECT EXISTS
+    ( SELECT 1 FROM information_schema.schemata
+      WHERE schema_name = $1
+    ) |] (Identity schemaName) False
+
+
+doesTableExist :: (MonadTx m) => SchemaName -> TableName -> m Bool
+doesTableExist schemaName tableName =
+  liftTx $ (runIdentity . Q.getRow) <$> Q.withQE defaultTxErrorHandler [Q.sql|
+    SELECT EXISTS
+    ( SELECT 1 FROM pg_tables
+      WHERE schemaname = $1 AND tablename = $2
+    ) |] (schemaName, tableName) False
+
+isExtensionAvailable :: (MonadTx m) => SchemaName -> m Bool
+isExtensionAvailable schemaName =
+  liftTx $ (runIdentity . Q.getRow) <$> Q.withQE defaultTxErrorHandler [Q.sql|
+    SELECT EXISTS
+    ( SELECT 1 FROM pg_catalog.pg_available_extensions
+      WHERE name = $1
+    ) |] (Identity schemaName) False
+
+isCatalogMigrateRequired
+  :: forall m
+   . ( MonadIO m
+     , MonadTx m
+     )
+  => m Bool
+isCatalogMigrateRequired = do
+  doesSchemaExist (SchemaName "hdb_catalog") >>= \case
+    False -> pure True
+    True  -> doesTableExist (SchemaName "hdb_catalog") (TableName "hdb_version") >>= \case
+      False -> pure True
+      True  -> do
+        dbCatalogVersion <- liftTx getCatalogVersion
+        pure $ dbCatalogVersion /= latestCatalogVersionString
```

Comment:
Instead of having a predicate that this module exports to guard calls to `migrateCatalog`, I think it would be nicer to move the logic into `migrateCatalog` itself. Then we don’t have to pollute the call site with extra logic that really only this module should have any reason to care about.